### ISSUE TITLE
Use Pigeon for isolate communication

### ISSUE THREAD

**Issue by matthew-carroll** [flutter/flutter#118767]
Pigeon is an official Flutter package that generates communication implementation to cross from Flutter's Dart land into platform land.

There's a similar challenge when communicating between isolates. We might like to call methods and receive return values across isolates, as if we were executing any other asynchronous method. Doing this manually is very tedious. Every method call needs an implementation on the sender side, a message encoder/decoder, and another version of the method on the callee side.

Does Pigeon contain tools that could be used in their current form to generate comms for isolates? If not, can Pigeon add that machinery?

@gaaclarke

**Comment by darshankawar** [comment]
@matthew-carroll 
I think it sounds similar to https://github.com/flutter/flutter/issues/72030, so can you confirm and likely follow-up in that issue ?

**Comment by stuartmorgan** [comment]
#72030 is about shared memory rather than a method-calling system, so this request is different.

> Does Pigeon contain tools that could be used in their current form to generate comms for isolates?

No, the generated code wraps platform channels.

We have currently no plans to expand Pigeon beyond the scope of Dart <-> host communication, but I'll leave this open for future consideration.

**Comment by gaaclarke** [comment]
@matthew-carroll Yep, on my radar but no definite plans.  When developing Flutter apps there are 3 boundaries that people run into: SendPort, Platform Channels, FFI.  I would like it if Pigeon could be used to abstract them all away.  There is another issue for generating FFI code which is the other side of this.

**Comment by matthew-carroll** [comment]
If there's enough demand from companies on the ports and FFI side, this could be a possible Flutter Bounty Hunters project. I'm sure you'll want to keep Pigeon with the Flutter team, so it would be an additional project, but if the funding is there, we'll build it.

Or, if the funding is there, I could have the Flutter Bounty Hunters work on Pigeon, directly.

Food for thought.
